### PR TITLE
Improve bash compile script

### PR DESCRIPTION
**Pull request by leuc**
set CC for custom GCC path
set MAX_JOBS=1 for low memory builds
be more portable (it's still bash not POSIX shell. yet)
do requirement checks first, fail early
only auto-install python requirements

also addressing the issues raised in https://github.com/intel/intel-extension-for-pytorch/issues/329

### CODE REVIEW

**Comment by leuc** [comment]
The sanity check at the bottom of the script only works for me when libgsm pre-loaded

```
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgsm.so.1 python -c "import torch; ...
```

or it fails with 
```
python3.10/site-packages/torchaudio/lib/libtorchaudio.so: undefined symbol: gsm_create
```
Is there something wrong the way torch audio is build?

File: scripts/compile_bundle.sh [hunk 1]
```diff
@@ -1,5 +1,6 @@
-#!/bin/bash
-set -x
+#!/usr/bin/env bash
+set -veo pipefail
+# can not use -u, because external env/vars.sh source'ing
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:3]: would you elaborate more on using `/usr/bin/env bash` and `set -veo pipefail`?

File: scripts/compile_bundle.sh [hunk 2]
```diff
@@ -26,35 +27,42 @@ if [ ! -f ${DPCPP_ENV} ]; then
     echo "DPC++ compiler environment ${DPCPP_ENV} doesn't seem to exist."
     exit 2
 fi
+source ${DPCPP_ENV}
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:30]: dpcpp_env and mkl_env is only needed when compiling ipex.

File: scripts/compile_bundle.sh [hunk 3]
```diff
@@ -153,31 +156,20 @@ unset LLVM_DIR
...
-conda install -y libpng jpeg
 python setup.py clean
 python setup.py bdist_wheel 2>&1 | tee build.log
 python -m pip install --force-reinstall --no-deps dist/*.whl
-python -m pip install Pillow
-git config --global --unset safe.directory
 #  TorchAudio
 cd ../audio
-git config --global --add safe.directory `pwd`
-conda install -y bzip2
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:169]: why these conda install commands are removed?

Comment (leuc) [reply scripts/compile_bundle.sh:169]: not everybody is using conda and this script should focus on compiling the whl files. it fails early if the requirements are not met and the user can take the needed actions to fulfill them. it's simply more portable.

File: scripts/compile_bundle.sh (hunk 1 above)
Comment (leuc) [reply scripts/compile_bundle.sh:3]: These are common best practice for bash scripts. -e and -o pipefail stop the script on any errors. `set -u` is highly recommended, to fail on any undefined variables. This avoids undefined and unexpected behaviors, especially here were ENV vars change the build options. `/usr/bin//env bash` ensure a working script on OS's where bash is not under /bin/bash

i might add a change to enable / disable set -u / set +u only around the source .../env/vars.sh includes 

For CI/CD it would actually be wise to use a POSIX shell script, not bash, because the smaller Alpine docker images don't even have bash. However, given this script depends on oneAPI anyway and it's supported distribution, bash should be fine.

File: scripts/compile_bundle.sh (hunk 2 above)
Comment (leuc) [reply scripts/compile_bundle.sh:30]: Yes, i wondered if it would have any side effect on the other build steps? 

The reason here was to fail early on external dependencies. 
(It's really annoying to have a 2h build fail in the middle, especially when it always re-builds everything)

I also noticed that the ENV variables might get mixed up, when the script runs multiple times. It might be wise to reset all used ENV vars at the beginning? Should i add that?

Comment (jingxu10) [reply scripts/compile_bundle.sh:30]: I'm not very clear why source command would fail. Existence of the source target is checked before it. Did you experience failures caused by source before or this is just in case?
Our internal compilation of pytorch and torch vision doesn't have dpcpp env and oneapi mkl activated. Sourcing them afterwards is to align with our internal process.
When this script is executed with `bash compile_bundle.sh`, all environment variables set inside this bash script will be discarded when the execution finishes. Why ENV variables get mixed up when the script runs for multiple times?

File: scripts/compile_bundle.sh [hunk 4]
```diff
@@ -26,35 +27,42 @@ if [ ! -f ${DPCPP_ENV} ]; then
...
-fi
-which patch > /dev/null 2>&1
-if [[ $? -ne 0 ]]; then
-    echo "Error: linux command \"patch\" not found."
-    exit 6
-fi
-which pkg-config > /dev/null 2>&1
-if [[ $? -ne 0 ]]; then
-    echo "Error: linux command \"pkg-config\" not found."
+for APP in python git patch pkg-config nproc bzip2; do
```
[the remark this answers is not in the document]
Comment (jingxu10) [reply scripts/compile_bundle.sh:40]: Would you add `gcc` and `g++` check here?

File: scripts/compile_bundle.sh [hunk 5]
```diff
@@ -26,35 +27,42 @@ if [ ! -f ${DPCPP_ENV} ]; then
...
+done
+
+# Check existance of required libs
+for LIB_NAME in libpng libjpeg; do
+    pkg-config --exists $LIB_NAME || (echo "Error: \"${LIB_NAME}\" not found in pkg-config." ; exit 5)
+done
+
+# set CC if not already defined
+if [ -z "${CC-}" ]; then
+    export CC="$(command -v gcc)" || (echo "Error: gcc not found" ; exit 6)
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:51]: Why setting CC?

File: scripts/compile_bundle.sh [hunk 6]
```diff
@@ -26,35 +27,42 @@ if [ ! -f ${DPCPP_ENV} ]; then
...
+fi
+
+if [ $($CC -dumpversion) -ne 11 ]; then
+    echo "Error: GCC version 11 required"
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:55]: Would you dump gcc version out to notify users the active gcc version rather than check and exit if it is not gcc 11? It is highly recommended to check installation guide first for requirements, including driver version, oneAPI basekit version, etc.

File: scripts/compile_bundle.sh [hunk 7]
```diff
@@ -99,44 +107,39 @@ git checkout ${VER_IPEX}
 git submodule sync
 git submodule update --init --recursive
 
-# Install the basic dependency cmake
-python -m pip install cmake
-
 # Compile individual component
 #  LLVM
 cd ../llvm-project
-git config --global --add safe.directory `pwd`
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:108]: When compiling in a docker container with root account, this `git config` command is required. Would you add a UID check around it?
```
if [ ${UID} -eq 0 ]; then
    git config --global --add safe.directory `pwd`
fi
```

File: scripts/compile_bundle.sh [hunk 8]
```diff
@@ -99,44 +107,39 @@ git checkout ${VER_IPEX}
...
-if [[ ${CONDA} -eq 0 ]]; then
-    export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}
+if [ -n "${CONDA_PREFIX-}" ]; then
+    export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(command -v conda))/../"}
+elif [ -n "${VIRTUAL_ENV-}" ]; then
+    export CMAKE_PREFIX_PATH=${VIRTUAL_ENV:-"$(dirname $(command -v python))/../"}
 else
-    export CMAKE_PREFIX_PATH=${VIRTUAL_ENV:-"$(dirname $(which python))/../"}
+    # TODO not building with conda or virtualenv, what should be set?
+    continue
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:142]: Did you try running this script in a non-conda python environment?

File: scripts/compile_bundle.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (leuc) [reply scripts/compile_bundle.sh:40]: yes, makes sense.

File: scripts/compile_bundle.sh (hunk 6 above)
Comment (leuc) [reply scripts/compile_bundle.sh:55]: i was adopting the changes from https://github.com/intel/intel-extension-for-pytorch/pull/331 as they reported issues with other gcc versions?

https://intel.github.io/intel-extension-for-pytorch/cpu/latest/tutorials/installation.html#system-requirements

OK, installation guide: "**Recommend** using GCC 11", so i would adopt this into a "warning" instead.

File: scripts/compile_bundle.sh (hunk 7 above)
Comment (leuc) [reply scripts/compile_bundle.sh:108]: Thanks for clarifying why this is required. This was failing on me when re-running failed build attempts. I will figure out a robust solution.

Comment (jingxu10) [reply scripts/compile_bundle.sh:108]: Removing [safe] section contents in ~/.gitconfig should solve the failure that you encountered.

File: scripts/compile_bundle.sh (hunk 6 above)
Comment (jingxu10) [reply scripts/compile_bundle.sh:55]: Since checking gcc with a specific version in this script brings extra maintenance costs. I'm wondering if just printing the gcc version out could notify users to check the installation guide or not.

File: scripts/compile_bundle.sh (hunk 8 above)
Comment (leuc) [reply scripts/compile_bundle.sh:142]: Yes, my builds and tests were inside a python virtual env.

https://github.com/intel/intel-extension-for-pytorch/blob/master/docker/Dockerfile.compile

Could you clarify why conda is added to the build, when it is not actually used as environment? 
I don't see a creation of a Conda environment in the Dockerfile?

i assume, the script needs to support three types of builds?

1. within a conda env
2. within a python virtual env
3. within a docker container (this is basically the "default" case without any env, both as root and normal user)

What path needs to be added to `CMAKE_PREFIX_PATH` in the 3rd case?

Thanks for all the feedback and consideration of my PR.

Comment (jingxu10) [reply scripts/compile_bundle.sh:142]: It's OK to cover support of virtenv and normal python env.
For case 3, dependency python packages should be installed into /usr/lib? If so, no need to set CMAKE_PREFIX_PATH. The purpose of CMAKE_PREFIX_PATH is to make cmake able to find installed python packages in non-system search paths.

File: scripts/compile_bundle.sh (hunk 2 above)
Comment (leuc) [reply scripts/compile_bundle.sh:30]: the `source` command fails when `set -u` is active.

I will revert the change and see how to ensure a clean env variable scope. Even documented`bash compile_bundle.sh` is not a common practice. Everyone simply runs `./compile_bundle.sh` (including me).

File: scripts/compile_bundle.sh (hunk 7 above)
Comment (leuc) [reply scripts/compile_bundle.sh:108]: After reviewing git's `safe.directory` option i would recommend to add `git config --global --add 'safe.directory=*'` to the `Dockerfile` that calls this script.

The issue isn't that git runs as root, but mixed directory ownership's caused by the way docker mounts storage. This article explains the issue:

https://www.kenmuse.com/blog/avoiding-dubious-ownership-in-dev-containers/

Triggering this issue inside docker depends on the actual docker setup. It would add a lot of complexity to this script to detect such scenario, since it's not simply "running as root" or "running inside docker".

It **might** also be possible to avoid the issue by switching the user 
https://docs.docker.com/engine/reference/builder/#user

In conclusion, i would recommend leaving these commands out of this script to avoid changing the users git config. It's a docker specific issue and should be solved there.

Comment (jingxu10) [reply scripts/compile_bundle.sh:108]: Looks good to me. Let's remove the `git config` from this script then.
Would you add a note about invocation of `git config --global --add 'safe.directory=*'` to [installation guide](https://github.com/intel/intel-extension-for-pytorch/blob/xpu-master/docs/tutorials/installation.md#compile-the-bundle-pytorch-torchvision-torchaudio-intel-extension-for-pytorch-with-script)?
Something like `If the user performs compilation is different than that cloned down the repo, you need to invoke git confit ......`

File: scripts/compile_bundle.sh (hunk 2 above)
Comment (leuc) [reply scripts/compile_bundle.sh]: torch audio would not build without mkl, so i moved the script includes above that

File: scripts/compile_bundle.sh [hunk 9]
```diff
@@ -153,31 +156,25 @@ unset LLVM_DIR
...
 python setup.py clean
 python setup.py bdist_wheel 2>&1 | tee build.log
 python -m pip install --force-reinstall --no-deps dist/*.whl
-python -m pip install Pillow
-git config --global --unset safe.directory
+# don't fail on external scripts
+set +uex
+source ${DPCPP_ENV}
+source ${ONEMKL_ENV}
+set -uex
```
Comment (jingxu10) [inline scripts/compile_bundle.sh:168]: Does torchaudio compile with mkl-static? if so, please move line 157 `python -m pip uninstall -y mkl-static mkl-include` to the end of audio compilation.

Comment (leuc) [reply scripts/compile_bundle.sh:168]: torch audio doesn't seem to build against python mkl-static? the mkl requirement seems to come from the `third_party/kaldi/` sub-module, not audio itself.

Setting LD_LIBRARY_PATH, MKLROOT or CMAKE_PREFIX_PATH to the path where python mkl-static is installed, doesn't seem to help. 

```
FAILED: torchaudio/csrc/libtorchaudio.so
[...]
/usr/bin/ld: cannot find -lmkl_intel_ilp64: No such file or directory
/usr/bin/ld: cannot find -lmkl_intel_thread: No such file or directory
/usr/bin/ld: cannot find -lmkl_core: No such file or directory
collect2: error: ld returned 1 exit status
```
not sure how to proceed here

Comment (leuc) [reply scripts/compile_bundle.sh:168]: mkl requirement come from the `third_party/kaldi`/ sub-module

i am using virtualenv, not sure if this happens with conda.

Comment (jingxu10) [reply scripts/compile_bundle.sh:168]: Let's source MKL before compiling torchaudio then.

**Comment by jingxu10** [comment]
Hi, any pending work you are on?

**Comment by leuc** [comment]
> Hi, any pending work you are on?

no, i think the current state should build the whl's on most platforms.

above mentioned libgsm issue and the documentation update should move into separate PR.

**Comment by jingxu10** [comment]
Got it.
Thanks for the confirmation.

**Comment by jingxu10** [comment]
Hi, @leuc please check xpu-master, if it looks good, please close this PR.
